controllers/awscluster_controller_unit_test.go

```diff
@@ -0,0 +1,632 @@
+				return ec2Svc
+			},
+			elbServiceFactory: func(elbScope scope.ELBScope) services.ELBInterface {
+				return elbSvc
+			},
+			networkServiceFactory: func(clusterScope scope.ClusterScope) services.NetworkInterface {
+				return networkSvc
+			},
+			securityGroupFactory: func(clusterScope scope.ClusterScope) services.SecurityGroupInterface {
+				return sgSvc
+			},
+			Recorder: recorder,
+		}
+		return csClient
+	}
+
+	teardown := func(t *testing.T, g *WithT) {
+		t.Helper()
+		mockCtrl.Finish()
+	}
+
+	t.Run("Reconciling an AWSCluster", func(t *testing.T) {
+		t.Run("reconcile success", func(t *testing.T) {
+			t.Run("should successfully create AWSCluster with Cluster Finalizer and LoadBalancerReady status true on AWSCluster", func(t *testing.T) {
+				g := NewWithT(t)
+				runningCluster := func(t *testing.T, g *WithT) {
+					t.Helper()
+					ec2Svc.EXPECT().ReconcileBastion().Return(nil)
+					elbSvc.EXPECT().ReconcileLoadbalancers().Return(nil)
+					networkSvc.EXPECT().ReconcileNetwork().Return(nil)
+					sgSvc.EXPECT().ReconcileSecurityGroups().Return(nil)
+				}
+
+				awsCluster := getAWSCluster("test", "test")
+				csClient := setup(t, g, &awsCluster)
+				defer teardown(t, g)
+				runningCluster(t, g)
+				cs, err := scope.NewClusterScope(
+					scope.ClusterScopeParams{
+						Client:     csClient,
+						Cluster:    &clusterv1.Cluster{},
+						AWSCluster: &awsCluster,
+					},
+				)
+				g.Expect(err).To(BeNil())
+				awsCluster.Status.Network.APIServerELB.DNSName = "www.google.com"
+				awsCluster.Status.Network.APIServerELB.AvailabilityZones = []string{"us-east-1a", "us-east-1b", "us-east-1c", "us-east-1d", "us-east-1e"}
+				cs.SetSubnets(infrav1.Subnets{
+					{
+						ID:               "private-subnet-1",
+						AvailabilityZone: "us-east-1b",
+						IsPublic:         false,
+					},
+					{
+						ID:               "private-subnet-2",
+						AvailabilityZone: "us-east-1a",
+						IsPublic:         false,
+					},
+					{
+						ID:               "private-subnet-3",
+						AvailabilityZone: "us-east-1c",
+						IsPublic:         false,
+					},
+					{
+						ID:               "private-subnet-4",
+						AvailabilityZone: "us-east-1d",
+						IsPublic:         false,
+					},
+					{
+						ID:               "private-subnet-5",
+						AvailabilityZone: "us-east-1e",
+						IsPublic:         false,
+					},
+				})
+				_, err = reconciler.reconcileNormal(cs)
+				g.Expect(err).To(BeNil())
+				expectAWSClusterConditions(g, cs.AWSCluster, []conditionAssertion{{infrav1.LoadBalancerReadyCondition, corev1.ConditionTrue, "", ""}})
+				g.Expect(awsCluster.GetFinalizers()).To(ContainElement(infrav1.ClusterFinalizer))
+			})
+		})
+		t.Run("reconcile failure", func(t *testing.T) {
+			expectedErr := errors.New("failed to get resource")
+			t.Run("should fail AWSCluster create with reconcile network failure", func(t *testing.T) {
+				g := NewWithT(t)
+				awsCluster := getAWSCluster("test", "test")
```

Comment:
You are creating objects in same ns for multiple test cases, try to avoid it and create objects in separated ns

Reply:
answered that above

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3073#discussion_r796543557